File: urbanairship-core/src/main/java/com/urbanairship/iam/fullscreen/FullScreenAdapter.java
/* Copyright Airship and Contributors */

package com.urbanairship.iam.fullscreen;

import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;

import com.urbanairship.iam.DisplayHandler;
import com.urbanairship.iam.InAppMessage;
import com.urbanairship.iam.MediaDisplayAdapter;

/**
 * Full screen adapter.
 */
public class FullScreenAdapter extends MediaDisplayAdapter {

    /**
     * Default constructor.
     *
     * @param displayContent The display content.
     * @param message The in-app message.
     */
    protected FullScreenAdapter(@NonNull InAppMessage message, @NonNull FullScreenDisplayContent displayContent) {
        super(message, displayContent.getMedia());
    }

    /**
     * Creates a new full screen adapter.
     *
     * @param message The in-app message.
     * @return The full screen adapter.
     */
    @NonNull
    public static FullScreenAdapter newAdapter(@NonNull InAppMessage message) {
        FullScreenDisplayContent displayContent = message.getDisplayContent();
        if (displayContent == null) {
            throw new IllegalArgumentException("Invalid message for adapter: " + message);
        }

        return new FullScreenAdapter(message, displayContent);
    }

    @Override
    public void onDisplay(@NonNull Context context, @NonNull DisplayHandler displayHandler) {
        Intent intent = new Intent(context, FullScreenActivity.class)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                .putExtra(FullScreenActivity.DISPLAY_HANDLER_EXTRA_KEY, displayHandler)
                .putExtra(FullScreenActivity.IN_APP_MESSAGE_KEY, getMessage())
                .putExtra(FullScreenActivity.IN_APP_ASSETS, getAssets());

        context.startActivity(intent);
    }

}

File: urbanairship-automation/src/main/java/com/urbanairship/automation/rewrite/inappmessage/displayadapter/fullscreen/FullScreenAdapter.kt
/* Copyright Airship and Contributors */
package com.urbanairship.automation.rewrite.inappmessage.displayadapter.fullscreen

import android.content.Context
import android.content.Intent
import com.urbanairship.automation.rewrite.inappmessage.InAppActivityMonitor
import com.urbanairship.automation.rewrite.inappmessage.InAppMessageActivity
import com.urbanairship.automation.rewrite.inappmessage.analytics.InAppMessageAnalyticsInterface
import com.urbanairship.automation.rewrite.inappmessage.assets.AirshipCachedAssetsInterface
import com.urbanairship.automation.rewrite.inappmessage.content.InAppMessageDisplayContent
import com.urbanairship.automation.rewrite.inappmessage.displayadapter.DisplayAdapterInterface
import com.urbanairship.automation.rewrite.inappmessage.displayadapter.DisplayResult
import com.urbanairship.automation.rewrite.inappmessage.displayadapter.InAppMessageDisplayListener
import com.urbanairship.automation.rewrite.utils.ActiveTimer
import java.util.UUID
import kotlin.coroutines.Continuation
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext

/**
 * Full screen adapter.
 */
internal class FullScreenAdapter(
    private val displayContent: InAppMessageDisplayContent.FullscreenContent,
    private val assets: AirshipCachedAssetsInterface?,
    private val activityMonitor: InAppActivityMonitor
) : DisplayAdapterInterface {

    private var continuation: Continuation<DisplayResult>? = null

    override suspend fun display(
        context: Context,
        analytics: InAppMessageAnalyticsInterface
    ): DisplayResult {
        val token = UUID.randomUUID().toString()

        val displayListener = InAppMessageDisplayListener(
            analytics = analytics,
            timer = ActiveTimer(activityMonitor),
            onDismiss = {
                setDisplayListener(token, null)
                continuation?.resumeWith(Result.success(it))
            })

        setDisplayListener(token, displayListener)

        val intent =
            Intent(context, FullScreenActivity::class.java).setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                .putExtra(InAppMessageActivity.DISPLAY_LISTENER_TOKEN, token)
                .putExtra(InAppMessageActivity.DISPLAY_CONTENT, displayContent)
                .putExtra(InAppMessageActivity.IN_APP_ASSETS, assets)

        return withContext(Dispatchers.Main) {
            suspendCancellableCoroutine {
                continuation = it
                context.startActivity(intent)
            }
        }
    }

    override fun getIsReady(): Boolean = true
    override suspend fun waitForReady() {}

    private fun setDisplayListener(token: String, listener: InAppMessageDisplayListener?) {
        synchronized(messageToAnalytics) {
            if (listener == null) {
                messageToAnalytics.remove(token)
            } else {
                messageToAnalytics[token] = listener
            }
        }
    }

    companion object {
        private val messageToAnalytics = mutableMapOf<String, InAppMessageDisplayListener>()
        fun getListener(messageID: String): InAppMessageDisplayListener? {
            return synchronized(messageToAnalytics) { messageToAnalytics[messageID] }
        }
    }
}
